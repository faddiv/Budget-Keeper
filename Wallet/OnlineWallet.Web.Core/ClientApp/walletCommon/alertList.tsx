import * as React from "react";
import { bindActionCreators, compose } from "redux";
import { RouteComponentProps, withRouter } from "react-router";
import { connect } from "react-redux";
import { bind } from "bind-decorator";

import { AlertsActions } from "actions/alerts";
import { AlertMessage } from "reducers/alerts/alertsModel";
import { RootState } from "reducers";

export interface AlertListProps extends Partial<RouteComponentProps<void>> {
    alerts?: AlertMessage[];
    actions?: typeof AlertsActions;
}

export interface AlertListState {
}

export class AlertList2 extends React.Component<AlertListProps, AlertListState> {

    private navListener: () => void;

    constructor(props) {
        super(props);
        this.state = {
            items: []
        };
    }

    @bind
    private locationListener() {
        const { alerts, actions } = this.props;
        if (alerts.length > 0) {
            actions.dismissAllAlert();
        }
    }

    componentDidMount() {
        this.navListener = this.props.history.listen(this.locationListener);
    }

    componentWillUnmount() {
        this.navListener();
        this.navListener = null;
    }

    dismissAlert(alert: AlertMessage) {
        this.props.actions.dismissAlert(alert);
    }

    @bind
    renderAlert(alert: AlertMessage, index: number) {
        const css = "alert alert-" + alert.type;
        return (
            <div key={index} className={css} role="alert">
                {alert.message}
                <button type="button" className="close" aria-label="Close" onClick={() => this.dismissAlert(alert)}>
                    <span aria-hidden="true">&times;</span>
                </button>
            </div>
        );
    }

    render() {
        return (
            <div className="alerts">
                {this.props.alerts.map(this.renderAlert)}
            </div>
        );
    }
}

function mapStateToProps(state: RootState) {
    return {
        alerts: state.alerts
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(AlertsActions as any, dispatch) as typeof AlertsActions
    };
}

export const AlertList = compose(withRouter, connect(mapStateToProps, mapDispatchToProps))(AlertList2);
