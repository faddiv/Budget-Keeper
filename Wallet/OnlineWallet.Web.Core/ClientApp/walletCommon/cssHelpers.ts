export function className(...values: (string | boolean)[]): string {
    let show = true;
    let cssList: string[] = [];
    for (var value of values) {
        if (typeof (value) !== "string") {
            show = !!value;
            continue;
        }
        if (!show) {
            continue;
        }
        cssList.push(value);
    }
    if (cssList.length) {
        return cssList.join(" ");
    }
}