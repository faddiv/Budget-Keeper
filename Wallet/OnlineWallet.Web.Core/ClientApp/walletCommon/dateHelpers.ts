import * as moment from "moment";

export function toUTCDate(value: any): Date | undefined {
    if (value) {
        var date = moment(value);
        return new Date(Date.UTC(date.year(), date.month(), date.date()));
    }
}