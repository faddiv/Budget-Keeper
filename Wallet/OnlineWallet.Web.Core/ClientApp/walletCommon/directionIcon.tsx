import * as React from "react";
import { MoneyDirection } from "walletApi";

interface DirectionIconProps {
    direction: MoneyDirection;
}

export const DirectionIcon: React.SFC<DirectionIconProps> = ({ direction, ...rest }) => {
    return (
        <span className={directionCssClass(direction)}></span>
    );
};

function directionCssClass(direction: MoneyDirection) {
    switch (direction) {
        case -1:
            return "fa fa-minus text-danger";
        case 1:
            return "fa fa-plus text-success";
        default:
            return "fa fa-bookmark";
    }
}
