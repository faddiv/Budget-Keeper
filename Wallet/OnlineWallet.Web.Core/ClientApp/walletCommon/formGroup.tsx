import * as React from 'react';
import { ValidationStateElement } from 'helpers';
import { className } from 'react-ext';

type InputType = "text" | "button" | "submit" | "reset" | "radio" | "checkbox" | "color" |
    "date" | "datetime-local" | "email" | "month" | "number" | "range" | "search" | "tel" |
    "time" | "url" | "week";

interface FormGroupProps {
    id?: string,
    name: string,
    label: string,
    type?: InputType,
    value?: any,
    onChange?: () => void,
    autoComplete?: boolean
    validation?: ValidationStateElement
}

export const FormGroup: React.SFC<FormGroupProps> = ({ id, name, label, type, value, onChange, autoComplete, validation, ...rest }) => {
    id = id || name;
    return (
        <div className="form-group row">
            <label htmlFor={id} className="col-sm-2 col-form-label">{label}</label>
            <div className="col-sm-10">
                {rest.children ? rest.children : defaultInput(type, id, name, label, value, onChange, autoComplete, validation)}
            </div>
        </div>
    );
};

function defaultInput(type: string, id: string, name: string, label: string, value: any, onChange: () => void, autoComplete: boolean, validation: ValidationStateElement): React.ReactElement<any>[] {
    let elements: React.ReactElement<any>[] = [
        <input key={1} type={type} className={className("form-control", validation && validation.showError, "is-invalid")}
            id={id} name={name} placeholder={label} value={value} onChange={onChange} autoComplete={autoComplete ? null : "off"} />
    ];
    if (validation) {
        elements.push(
            <div key={2} className="invalid-feedback">
                {validation.message}
            </div>
        );
    }
    return elements;
}

FormGroup.defaultProps = {
    type: "text",
    onChange: () => { },
    autoComplete: false
};