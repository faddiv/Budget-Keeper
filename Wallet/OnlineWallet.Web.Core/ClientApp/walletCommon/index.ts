export * from "./alertList";
export * from "./directionIcon";
export * from "./editDelete";
export * from "./formGroup";
export * from "./saveCancel";
export * from "./pager";
export * from "./navbar";
export * from "./models";
export * from "./specialInputs";
export * from "./transactions-view";