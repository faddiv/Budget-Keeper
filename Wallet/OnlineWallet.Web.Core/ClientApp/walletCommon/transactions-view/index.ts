export * from "./transaction-table/transaction-table";