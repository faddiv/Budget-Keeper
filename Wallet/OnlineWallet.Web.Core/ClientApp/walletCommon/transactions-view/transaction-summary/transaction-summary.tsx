import * as React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { RouteComponentProps, withRouter } from "react-router";
import { bind } from "bind-decorator";

import { RootState } from "reducers";
import { TransactionViewModel } from "walletCommon";
import { TransactionSummaryActions, TransactionSummaryViewModel } from "actions/transactionsSummary";
import { MoneyDirection } from "walletApi";

export interface TransactionSummaryProps extends Partial<RouteComponentProps<void>> {
    transactionSummary?: TransactionViewModel[];
    actions?: typeof TransactionSummaryActions;
}

export interface TransactionSummaryState {
}

@connect(mapStateToProps, mapDispatchToProps)
@withRouter
export class TransactionSummary extends React.Component<TransactionSummaryProps, TransactionSummaryState> {

    private navListener: () => void;

    constructor(props) {
        super(props);
    }

    @bind
    private locationListener() {
        const { actions, transactionSummary } = this.props;
        if (transactionSummary.length > 0) {
            actions.transactionsSelected([]);
        }
    }

    componentDidMount() {
        this.navListener = this.props.history.listen(this.locationListener);
    }

    componentWillUnmount() {
        this.navListener();
        this.navListener = null;
    }

    @bind
    close() {
        this.props.actions.transactionsSelected([]);
    }

    render() {
        const { transactionSummary } = this.props;
        if (!transactionSummary || !transactionSummary.length) {
            return null;
        }
        const expenses = sumPrice(transactionSummary, MoneyDirection.Expense);
        const incomes = sumPrice(transactionSummary, MoneyDirection.Income);
        const plans = sumPrice(transactionSummary, MoneyDirection.Plan);
        return (
            <div className="transaction-summary card bg-light">
                <div className="crad-body">
                    <span>Expenses: {expenses}</span>&nbsp;
                    <span>Incomes: {incomes}</span>&nbsp;
                    <span>Plans: {plans}</span>
                    <button type="button" className="close" aria-label="Close" onClick={this.close}>
                        <span aria-hidden="true">&times;</span>
                    </button>
                </div>
            </div>
        );
    }
}

function sumPrice(items: TransactionSummaryViewModel, dir: MoneyDirection): number {
    let sum: number = 0;
    for (const element of items) {
        if (element.direction === dir) {
            sum += parseInt(element.price, 10);
        }
    }
    return sum;
}

function mapStateToProps(state: RootState) {
    return {
        transactionSummary: state.transactionSummary
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(TransactionSummaryActions as any, dispatch) as typeof TransactionSummaryActions
    };
}
