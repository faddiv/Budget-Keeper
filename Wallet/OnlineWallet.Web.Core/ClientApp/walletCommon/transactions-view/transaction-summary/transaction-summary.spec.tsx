import * as React from "react";
import * as H from "history";
import { shallow } from "enzyme";

import { TransactionSummary } from "./transaction-summary";
import { unwrap } from "helpers/testHelpers";
import { MoneyDirection } from "walletApi";
import { TransactionSummaryActions } from "actions/transactionsSummary";
import { TransactionViewModel } from "walletCommon";
import { createTransactionSummaryActionsMocks } from "actions/transactionsSummary.mocks";

describe("TransactionSummary", () => {
    const TransactionSummary2 = unwrap(TransactionSummary);
    let defaultHistory: H.History;
    let actions: typeof TransactionSummaryActions;

    beforeEach(() => {
        defaultHistory = H.createMemoryHistory();
        actions = createTransactionSummaryActionsMocks();
    });

    it("should sum incomes", () => {
        const summary: TransactionViewModel[] = [{
            direction: MoneyDirection.Income,
            price: "100"
        }, {
            direction: MoneyDirection.Income,
            price: "50"
        }, {
            direction: MoneyDirection.Expense,
            price: "50"
        }];

        const element = shallow(<TransactionSummary2 actions={actions} history={defaultHistory} transactionSummary={summary} />);

        expect(element).toBeDefined();
        expect(element.html()).toContain("Incomes: 150");
        expect(element).toMatchSnapshot();
    });

    it("should sum expenses", () => {
        const summary: TransactionViewModel[] = [{
            direction: MoneyDirection.Expense,
            price: "100"
        }, {
            direction: MoneyDirection.Expense,
            price: "50"
        }, {
            direction: MoneyDirection.Plan,
            price: "50"
        }];

        const element = shallow(<TransactionSummary2 actions={actions} history={defaultHistory} transactionSummary={summary} />);

        expect(element).toBeDefined();
        expect(element.html()).toContain("Expenses: 150");
        expect(element).toMatchSnapshot();
    });

    it("should sum plans", () => {
        const summary: TransactionViewModel[] = [{
            direction: MoneyDirection.Plan,
            price: "100"
        }, {
            direction: MoneyDirection.Plan,
            price: "50"
        }, {
            direction: MoneyDirection.Income,
            price: "50"
        }];

        const element = shallow(<TransactionSummary2 actions={actions} history={defaultHistory} transactionSummary={summary} />);

        expect(element).toBeDefined();
        expect(element.html()).toContain("Plans: 150");
        expect(element).toMatchSnapshot();
    });

    it("should call transactionsSelected with empty list when navigation occur", () => {
        const summary: TransactionViewModel[] = [{
            direction: MoneyDirection.Expense,
            price: "100"
        }, {
            direction: MoneyDirection.Plan,
            price: "50"
        }, {
            direction: MoneyDirection.Income,
            price: "50"
        }];

        const element = shallow(<TransactionSummary2 actions={actions} history={defaultHistory} transactionSummary={summary} />);
        defaultHistory.push("/otherLink");

        expect(element).toBeDefined();
        expect(actions.transactionsSelected).toBeCalledWith([]);
    });

    it("should render nothing with empty list", () => {
        const summary: TransactionViewModel[] = [];

        const element = shallow(<TransactionSummary2 actions={actions} history={defaultHistory} transactionSummary={summary} />);

        expect(element.html()).toBe(null);
    });
});
