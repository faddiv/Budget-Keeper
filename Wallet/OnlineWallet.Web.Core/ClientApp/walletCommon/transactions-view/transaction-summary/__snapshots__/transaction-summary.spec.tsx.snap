// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionSummary should sum expenses 1`] = `
<div
  className="transaction-summary card bg-light"
>
  <div
    className="crad-body"
  >
    <span>
      Expenses: 
      150
    </span>
     
    <span>
      Incomes: 
      0
    </span>
     
    <span>
      Plans: 
      50
    </span>
    <button
      aria-label="Close"
      className="close"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;

exports[`TransactionSummary should sum incomes 1`] = `
<div
  className="transaction-summary card bg-light"
>
  <div
    className="crad-body"
  >
    <span>
      Expenses: 
      50
    </span>
     
    <span>
      Incomes: 
      150
    </span>
     
    <span>
      Plans: 
      0
    </span>
    <button
      aria-label="Close"
      className="close"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;

exports[`TransactionSummary should sum plans 1`] = `
<div
  className="transaction-summary card bg-light"
>
  <div
    className="crad-body"
  >
    <span>
      Expenses: 
      0
    </span>
     
    <span>
      Incomes: 
      50
    </span>
     
    <span>
      Plans: 
      150
    </span>
    <button
      aria-label="Close"
      className="close"
      onClick={[Function]}
      type="button"
    >
      <span
        aria-hidden="true"
      >
        ×
      </span>
    </button>
  </div>
</div>
`;
